File: plugins/captcha-handler/lua/captcha_handler.lua
-- Lua-facing policy helper for plugins/captcha-handler.
--
-- This module mirrors the native plugin's host-mediated posture: it classifies
-- response evidence, selects an allowed handling plan, and never logs or
-- returns secret-bearing cookie/token values.
local M = {_version = "0.1.0"}

local function lower(value)
    return tostring(value or ""):lower()
end

local function header(headers, name)
    name = lower(name)
    for key, value in pairs(headers or {}) do
        if lower(key) == name then return tostring(value or "") end
    end
    return nil
end

local function add_signal(result, source, name, detail, confidence)
    result.signals[#result.signals + 1] = {
        source = source,
        name = name,
        detail = detail,
        confidence = confidence
    }
    if confidence > result.confidence then result.confidence = confidence end
    result.activated = result.confidence >= 0.55
end

local function classify(result)
    local captcha, rate_limit, waf = false, false, false
    for _, signal in ipairs(result.signals) do
        local text = lower((signal.name or "") .. " " .. (signal.detail or ""))
        captcha = captcha or text:find("captcha", 1, true) ~= nil or
            text:find("recaptcha", 1, true) ~= nil or
            text:find("hcaptcha", 1, true) ~= nil or
            text:find("turnstile", 1, true) ~= nil or
            text:find("human%-verification") ~= nil
        rate_limit = rate_limit or text:find("rate limit", 1, true) ~= nil or
            text:find("too many requests", 1, true) ~= nil
        waf = waf or text:find("cloudflare", 1, true) ~= nil or
            text:find("datadome", 1, true) ~= nil or
            text:find("distil", 1, true) ~= nil
    end
    if captcha then
        result.category = "captcha"
    elseif rate_limit then
        result.category = "rate-limit"
    elseif waf then
        result.category = "waf-challenge"
    elseif #result.signals > 0 then
        result.category = "anti-bot"
    end
end

function M.inspect_response(method, url, response)
    local result = {
        activated = false,
        category = "",
        url = (response and response.final_url ~= "" and response.final_url) or url or "",
        status = tonumber(response and response.status) or 0,
        confidence = 0,
        signals = {},
        server_confirmed = false,
        diagnostic = "no challenge evidence in HTTP response"
    }
    local body = lower(response and response.body or "")
    local headers = (response and response.headers) or {}

    if result.status == 403 then
        add_signal(result, "http", "forbidden",
            "HTTP 403 can indicate a WAF or bot challenge", 0.55)
    elseif result.status == 429 then
        add_signal(result, "http", "rate-limit", "HTTP 429 Too Many Requests", 0.70)
    elseif result.status == 503 then
        add_signal(result, "http", "service-unavailable",
            "HTTP 503 can indicate a temporary challenge page", 0.45)
    end

    if lower(header(headers, "cf-mitigated")) == "challenge" then
        add_signal(result, "header", "cloudflare-challenge",
            "cf-mitigated header reports challenge", 0.95)
    end
    if header(headers, "x-datadome") or header(headers, "x-distil-cs") then
        add_signal(result, "header", "bot-protection-header",
            "bot protection response header", 0.85)
    end

    if body:find("g%-recaptcha") or body:find("google.com/recaptcha", 1, true) then
        add_signal(result, "response-body", "recaptcha", "Google reCAPTCHA marker", 0.90)
    end
    if body:find("h%-captcha") or body:find("hcaptcha.com", 1, true) then
        add_signal(result, "response-body", "hcaptcha", "hCaptcha marker", 0.90)
    end
    if body:find("cf%-turnstile") or body:find("challenges.cloudflare.com", 1, true) then
        add_signal(result, "response-body", "turnstile", "Cloudflare Turnstile marker", 0.90)
    end
    if body:find("human verification", 1, true) or
       body:find("verify you are human", 1, true) or
       body:find("verification required", 1, true) then
        add_signal(result, "response-body", "human-verification",
            "human verification challenge text", 0.75)
    elseif body:find("captcha", 1, true) then
        add_signal(result, "response-body", "captcha-text",
            "page text mentions captcha", 0.70)
    end

    classify(result)
    if result.activated then
        for _, signal in ipairs(result.signals) do
            local evidence = lower((signal.name or "") .. " " .. (signal.detail or ""))
            if signal.source == "header" or result.status == 403 or result.status == 429 or
               evidence:find("recaptcha", 1, true) or
               evidence:find("hcaptcha", 1, true) or
               evidence:find("turnstile", 1, true) or
               evidence:find("human%-verification") then
                result.server_confirmed = true
                break
            end
        end
        result.diagnostic = result.server_confirmed and
            "server response contains anti-bot challenge evidence" or
            "challenge is heuristic and lacks a server confirmation signal"
    end
    return result
end

function M.select_handling_method(result, inputs, options)
    inputs = inputs or {}
    options = options or {}
    if not result or not result.activated then return "abort-and-report" end
    if inputs.has_clearance_cookie and options.allow_cookie_session_reuse ~= false then
        return "cookie-session-reuse"
    end
    if inputs.has_pre_solved_token and options.allow_pre_solved_token ~= false then
        return "pre-solved-token"
    end
    if result.category == "rate-limit" and options.allow_wait_for_clearance ~= false then
        return "wait-for-clearance"
    end
    if options.allow_manual_user_prompt then return "manual-user-prompt" end
    return "abort-and-report"
end

return M

File: plugins/captcha-handler/src/captcha_handler.cpp
#include "prowsetk/plugins/captcha_handler.hpp"

#include <utility>

namespace prowsetk::plugins::captcha_handler {
namespace {

HandlingPlan plan(HandlingMethod method, bool available, std::string name,
                  std::string description, std::string configuration_key,
                  std::string security_note) {
    return HandlingPlan{method,
                        available,
                        std::move(name),
                        std::move(description),
                        std::move(configuration_key),
                        std::move(security_note)};
}

}  // namespace

const char* to_string(HandlingMethod method) noexcept {
    switch (method) {
        case HandlingMethod::ManualUserPrompt: return "manual-user-prompt";
        case HandlingMethod::ExternalSolverApi: return "external-solver-api";
        case HandlingMethod::WebhookDispatch: return "webhook-dispatch";
        case HandlingMethod::LuaCallback: return "lua-callback";
        case HandlingMethod::PreSolvedToken: return "pre-solved-token";
        case HandlingMethod::CookieSessionReuse: return "cookie-session-reuse";
        case HandlingMethod::WaitForClearance: return "wait-for-clearance";
        case HandlingMethod::AbortAndReport: return "abort-and-report";
    }
    return "abort-and-report";
}

std::vector<HandlingPlan> available_methods(
    const CaptchaHandlerOptions& options) {
    return {
        plan(HandlingMethod::ManualUserPrompt, options.allow_manual_user_prompt,
             "Manual user prompt",
             "pause automation and ask the embedding application to obtain a token",
             "manual_prompt", "challenge content is not logged"),
        plan(HandlingMethod::ExternalSolverApi, options.allow_external_solver,
             "External solver API",
             "delegate challenge metadata to a configured solver service",
             "solver_api", "API keys and returned tokens must be redacted"),
        plan(HandlingMethod::WebhookDispatch, options.allow_webhook,
             "Webhook dispatch",
             "send a host-mediated webhook for out-of-process solving",
             "webhook_url", "webhook payloads omit cookies by default"),
        plan(HandlingMethod::LuaCallback, options.allow_lua_callback,
             "Lua callback",
             "call a Lua extension callback with heuristic challenge metadata",
             "lua_callback", "Lua receives no raw native handles"),
        plan(HandlingMethod::PreSolvedToken, options.allow_pre_solved_token,
             "Pre-solved token",
             "inject a token supplied by the caller before retrying",
             "pre_solved_token", "token values are treated as secrets"),
        plan(HandlingMethod::CookieSessionReuse,
             options.allow_cookie_session_reuse, "Cookie session reuse",
             "reuse a caller-provided clearance cookie or session profile",
             "clearance_cookie", "cookies remain under the storage policy"),
        plan(HandlingMethod::WaitForClearance, options.allow_wait_for_clearance,
             "Wait for clearance",
             "retry later when a temporary WAF or rate-limit challenge clears",
             "retry_policy", "bounded retries only; no wall-clock sleeps in tests"),
        plan(HandlingMethod::AbortAndReport, true, "Abort and report",
             "stop the workflow and return challenge provenance to the caller",
             "abort", "default fallback when no solver is configured"),
    };
}

CaptchaHandlerResult inspect_document(const Document& document,
                                      const CaptchaHandlerOptions& options) {
    AntiBotDetector detector;
    CaptchaHandlerResult result;
    result.detection = detector.inspect_document(document);
    result.methods = available_methods(options);
    return result;
}

}  // namespace prowsetk::plugins::captcha_handler
